File: q0316/q316.java
class Solution {
    public String removeDuplicateLetters(String s) {
        Map<Character, Integer> indexMap = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            indexMap.put(s.charAt(i), i);
        }

        Deque<Character> stack = new ArrayDeque<>();
        Set<Character> set = new HashSet<>();
        for (int i = 0; i < s.length(); i++) {
            char currChar = s.charAt(i);
            // only check and try to push the currChar if it is not already on the stack
            if (!set.contains(currChar)) { // if (!stack.contains(currChar)) {
                // if the incoming character is smaller than the character currently on the stack, and 
                // if the character on the stack could appear later, pop that character out as it can be added later
                // use a while loop for this as we want to repeatedly do so until what is on the stack absolutely has to be there
                while (!stack.isEmpty() && currChar < stack.peek() && indexMap.get(stack.peek()) > i) {
                    set.remove(stack.pop()); // stack.pop();
                }
                stack.push(currChar);
                set.add(currChar);
            }
        }

        StringBuilder sb = new StringBuilder();
        while (!stack.isEmpty()) sb.append(stack.pop());
        return sb.reverse().toString();
    }
}

File: q0316/q316.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        final_index_map = {char: i for i, char in enumerate(s)}
        visited = set()
        stack = []
        for i, char in enumerate(s):
            if char not in visited:
                while stack and stack[-1] > char and i < final_index_map[stack[-1]]:
                    visited.remove(stack[-1])
                    stack.pop()
                stack.append(char)
                visited.add(char)
        return "".join(stack)
